docs: Documentation refresh

rthill91 opened the pull request:
- restore missing 'howto_publish' document
- create installation guide
- update index to provide clearer navigation

toabctl reviewed (changes requested):
CI is unhappy:
```
/home/runner/work/awspub/awspub/docs/installation.rst: WARNING: document isn't included in any toctree
```

rthill91 commented:
> CI is unhappy:
> 
> ```
> /home/runner/work/awspub/awspub/docs/installation.rst: WARNING: document isn't included in any toctree
> ```

fixed. Had it in there originally, not sure how/when I dropped it

k-dimple commented on docs/howto_publish.rst:
```diff
@@ -0,0 +1,252 @@
+
+To publish a VMDK file as an image (AMI) on AWS EC2 with `awspub`,
+a config file is required.
+A basic config file looks like:
```
```suggestion
A basic config file (say ``config.yaml``) looks like:
```

k-dimple commented on docs/howto_publish.rst:
```diff
@@ -0,0 +1,252 @@
+.. literalinclude:: config-samples/config-with-parameters.yaml
+   :language: yaml
+
+In this config, the identifier which will be replaced is `$serial`. The mapping
```
```suggestion
In the config file shown above, the identifier `$serial` will be replaced with a value that is defined in another YAML file.
```

k-dimple commented on docs/howto_publish.rst:
```diff
@@ -0,0 +1,252 @@
+   :language: yaml
+
+In this config, the identifier which will be replaced is `$serial`. The mapping
+that is used to replace an identifier with a actual value is also a yaml file
```
Delete this line.

k-dimple commented on docs/howto_publish.rst:
```diff
@@ -0,0 +1,252 @@
+
+In this config, the identifier which will be replaced is `$serial`. The mapping
+that is used to replace an identifier with a actual value is also a yaml file
+which contains a mapping structure (dict in Python) that maps identifiers
```
```suggestion
This YAML file (say ``config-mapping.yaml``) contains a mapping structure (dict in Python) that maps identifiers
```

k-dimple commented on docs/howto_publish.rst:
```diff
@@ -0,0 +1,252 @@
+.. literalinclude:: config-samples/config-with-parameters.yaml.mapping
+   :language: yaml
+
+Using both together now with `awspub`:
```
```suggestion
Using both these config files, the command for `awspub` becomes:
```

k-dimple commented on docs/howto_publish.rst:
```diff
@@ -0,0 +1,252 @@
+~~~~~~~~~~~~
+
+There might be cases were the different commands (e.g. `awspub create` or `awspub publish`)
+should only be applied for a subset of the defined images. That's possible with the `group`
```
```suggestion
should only be applied on a subset of the defined images. That's possible with the `groups`
```

k-dimple replied:
:) I changed it to 'groups' because the keyword in the config file is that!

rthill91 replied:
OOH you're right! it's just the argument flag that's singular, great catch!

k-dimple commented on docs/howto_publish.rst:
```diff
@@ -0,0 +1,252 @@
+   :language: yaml
+
+Now the different `awspub` commands do have a `--group` parameter to filter which
+images this command should operate on:
```
Change this statement to:
Use the --group parameter to filter the images that the `awspub` command should operate on:

k-dimple commented on docs/howto_publish.rst:
```diff
@@ -0,0 +1,252 @@
+    }
+  }
+
+So the first command only works with images defined in `group1` while the second command only with
```
```suggestion
The first command is applied only to images defined in `group1`, while the second one is applied only to
```

k-dimple commented on docs/howto_publish.rst:
```diff
@@ -0,0 +1,252 @@
+.. literalinclude:: config-samples/config-minimal-public.yaml
+   :language: yaml
+
+The the image needs to be created and published:
```
```suggestion
The image needs to be created and then published:
```

k-dimple commented on docs/howto_publish.rst:
```diff
@@ -0,0 +1,252 @@
+AWS Marketplace
+~~~~~~~~~~~~~~~
+
+It's possible to publish to `AWS Marketplace <https://docs.aws.amazon.com/marketplace/latest/userguide/user-guide-for-sellers.html>`_ if a entity of a `Single-AMI product <https://docs.aws.amazon.com/marketplace/latest/userguide/ami-single-ami-products.html>`_ already exists, an access role ARN is available and an AMI exists in the `us-east-1` region. 
```
```suggestion
It's possible to publish to `AWS Marketplace <https://docs.aws.amazon.com/marketplace/latest/userguide/user-guide-for-sellers.html>`_ if an entity of a `Single-AMI product <https://docs.aws.amazon.com/marketplace/latest/userguide/ami-single-ami-products.html>`_ already exists, an access role ARN is available and an AMI exists in the `us-east-1` region. 
```

k-dimple commented on docs/howto_publish.rst:
```diff
@@ -0,0 +1,252 @@
+~~~~~~~~~~~~~~~~~~~
+
+It's possible to push information about published images to the `SSM Parameter Store <https://docs.aws.amazon.com/systems-manager/latest/userguide/systems-manager-parameter-store.html>`_. That's
+useful e.g. to have a common way to get the latest image Id on different
```
```suggestion
useful e.g. to have a common way to get the latest image ID in different
```

k-dimple commented on docs/howto_publish.rst:
```diff
@@ -0,0 +1,252 @@
+.. literalinclude:: config-samples/config-minimal-ssm.yaml
+   :language: yaml
+
+with a corresponding mapping file:
```
```suggestion
along with a corresponding mapping file:
```

k-dimple commented on docs/howto_publish.rst:
```diff
@@ -0,0 +1,252 @@
+.. literalinclude:: config-samples/config-minimal-ssm.yaml.mapping
+   :language: yaml
+
+Create the image and the `public` command will also push information to the parameter store:
```
```suggestion
Create the image and use the `public` command to publish the image and also push information to the parameter store:
```

k-dimple commented on docs/howto_publish.rst:
```diff
@@ -0,0 +1,252 @@
+Resource tags
+~~~~~~~~~~~~~
+
+The different resources (S3 objects, snapshots and AMIs) can have tags.
```
```suggestion
The different AWS resources (S3 objects, snapshots and AMIs) can have tags associated with them.
```

k-dimple commented on docs/installation.rst:
```diff
@@ -0,0 +1,46 @@
+Install awspub using snap
+-------------------------
+
+``awspub`` is available in the `Snapstore`_, and it can be installed using:
```
Let's use the same formatting of awspub throughout i.e. either `awspub` or ``awspub``

k-dimple commented on index.rst:
```diff
@@ -5,18 +5,51 @@ Welcome to awspub's documentation!
+
+----
+
+In this documentation
```
This should ideally be just two sections: How-to guides (containing installation and howto_publish) and Reference (containing architecture, reference_index and reference_config). For this you'll need to change the directory structure, and include index files for the two sub-folders corresponding to How-to guides and Reference. The toctree will also have to be corrected appropriately.

Otherwise, at least change the organisation to:

.. grid:: 1 1 2 2
   :padding: 0

   .. grid-item:: :doc:`Installation <docs/installation>`

       **Installation** covers how to setup and install awspub     
       
   .. grid-item:: :doc:`How to publish <docs/howto_publish>`

       **Publish** covers how to publish images to AWS

.. grid:: 1 1 2 2
   :padding: 0

   .. grid-item:: :doc:`Architecture <docs/architecture>`

       **Architecture** covers architectural decisions and considerations

   .. grid-item:: :doc:`Reference <docs/reference_index>`

       **Reference** Auto-generated reference pages for all classes and methods

       **ConfigModel Reference** Auto-generated reference pages for backing config models

k-dimple commented on index.rst:
```diff
@@ -5,18 +5,51 @@ Welcome to awspub's documentation!
+
+   .. grid-item:: :doc:`ConfigModel Reference <docs/reference_config>`
+
+       **ConfigModel Reference** Auto-generated reference pages for backing config models
```
what does 'backing config models' mean?

rthill91 commented on .custom_wordlist.txt:
```diff
@@ -1,3 +1,4 @@
+AMI
```
I can fix the yaml instance, but I think we'll require the lowercase `ssm`. It's used in reference to a config parameter (`ssm_parameter`) and I wouldn't want to change the casing there.

k-dimple replied:
Usually in such cases where it is a command / parameter / file name etc, I tend to put them within double backticks. That way they have a consistent format and are also not a problem from the spelling check point of view.

k-dimple commented:
Except for the 'groups' thing it looks pretty good to me. I'll make some minor style changes later (mainly to the new index pages).
